charmhelpers/contrib/network/ovs/__init__.py

```diff
@@ -83,36 +84,171 @@ def get_bridges_and_ports_map():
+        have a more pleasant data structure.
+
+        0: http://www.openvswitch.org/ovs-vswitchd.conf.db.5.pdf
+        1: https://tools.ietf.org/html/rfc7047
+        2: http://docs.openvswitch.org/en/latest/topics/integration/
+        3: https://github.com/openvswitch/ovs/blob/
+               20dac08fdcce4b7fda1d07add3b346aa9751cfbc/
+                   lib/db-ctl-base.c#L189-L215
+    :type data: Optional[Dict[str,Union[str,Dict[str,str]]]]
+    :param table: Name of table to operate on
+    :type table: str
+    :param entity: Name of entity to operate on
+    :type entity: str
+    :returns: '--' separated ``ovs-vsctl set`` commands
+    :rtype: Iterator[Tuple[str, str, str, str, str]]
+    """
+    for (k, v) in data.items():
+        if isinstance(v, dict):
+            entries = {
+                '{}:{}'.format(k, dk): dv for (dk, dv) in v.items()}
+        else:
+            entries = {k: v}
+        for (colk, colv) in entries.items():
+            yield ('--', 'set', table, entity, '{}={}'.format(colk, colv))
+
+
+def add_bridge(name, datapath_type=None, brdata=None, exclusive=False):
+    """Add the named bridge to openvswitch and set/update bridge data for it
+
+    :param name: Name of bridge to create
+    :type name: str
+    :param datapath_type: Add datapath_type to bridge (DEPRECATED, use brdata)
+    :type datapath_type: Optional[str]
+    :param brdata: Additional data to attach to bridge
+        The keys in the brdata dictionary map directly to column names in the
+        OpenvSwitch bridge table as defined in DB-SCHEMA [0] referenced in
+        RFC 7047 [1]
+
+        There are some established conventions for keys in the external-ids
+        column of various tables, consult the OVS Integration Guide [2] for
+        more details.
+
+        NOTE(fnordahl): Technically the ``external-ids`` column is called
+        ``external_ids`` (with an underscore) and we rely on ``ovs-vsctl``'s
+        behaviour of transforming dashes to underscores for us [3] so we can
+        have a more pleasant data structure.
+
+        0: http://www.openvswitch.org/ovs-vswitchd.conf.db.5.pdf
+        1: https://tools.ietf.org/html/rfc7047
+        2: http://docs.openvswitch.org/en/latest/topics/integration/
+        3: https://github.com/openvswitch/ovs/blob/
+               20dac08fdcce4b7fda1d07add3b346aa9751cfbc/
+                   lib/db-ctl-base.c#L189-L215
+    :type brdata: Optional[Dict[str,Union[str,Dict[str,str]]]]
+    :param exclusive: If True, raise exception if bridge exists
+    :type exclusive: bool
+    :raises: subprocess.CalledProcessError
+    """
     log('Creating bridge {}'.format(name))
-    cmd = ["ovs-vsctl", "--", "--may-exist", "add-br", name]
+    cmd = ['ovs-vsctl', '--']
+    if not exclusive:
+        cmd.append('--may-exist')
+    cmd.extend(('add-br', name))
+    if brdata:
+        for setcmd in _dict_to_vsctl_set(brdata, 'bridge', name):
+            cmd.extend(setcmd)
```

Comment:
Could we maybe add a log message indicating that the param is deprecated when it  is used.  This will light up in charms that still use it providing an opportunity to remove the paramter at some point?

Reply:
you're talking about `datapath_type`, right? I agree